File: julia/gc.jl
#!/usr/bin/env julia

function main()
    values = zeros(UInt64, 256)
    values[UInt8('A')+1] = values[UInt8('T')+1] = 1
    values[UInt8('G')+1] = values[UInt8('C')+1] = 1 << 32
    
    input = open("chry_multiplied.fa")
    buflen = 4096
    buf = zeros(UInt8, buflen)

    n = readbytes!(input, buf, buflen)

    count = UInt64(0)
    mask = UInt64(0)
    while n > 0
        if n < buflen
            resize!(buf, n)
        end

        for c in buf
            if c == UInt8('\n')
                mask = ~UInt64(0)
            elseif c == UInt8('>')
                mask = UInt64(0)
            else
                count += mask & @inbounds values[c+1]
            end
        end

        n = readbytes!(input, buf, buflen)
    end

    at_count = count & 0xffffffff
    gc_count = count >> 32

    gc = gc_count / (gc_count + at_count)
    println(gc)
end

main()

File: python/gc.py
#!/usr/bin/env python
import re
import string
 
def main():
    file = open("chry_multiplied.fa","r")
    a = 0
    t = 0
    g = 0
    c = 0
    for line in file:
        if line[0] != '>' and len(line) != line.count("N") + 1:
            g += line.count("G")
            c += line.count("C")
            a += line.count("A")
            t += line.count("T")
    totalBaseCount = a + t + c + g
    gcCount = g + c
    gcFraction = float(gcCount) / totalBaseCount
    print( gcFraction * 100 )
 
if __name__ == '__main__':
    main()
